src/devices/interlogix.c

```diff
@@ -0,0 +1,345 @@
+{
+    unsigned char *b = (unsigned char*) ptr;
+    unsigned char byte;
+    int i, j;
+
+    for (i=size-1;i>=0;i--)
+    {
+        for (j=7;j>=0;j--)
+        {
+            byte = (b[i] >> j) & 1;
+            fprintf(stderr, "%u", byte);
+        }
+    }
+    fprintf(stderr, " ");
+}
+
+char * int2bin(uint8_t i)
+{
+    size_t bits = sizeof(uint8_t) * 8;
+
+    char * str = malloc(bits + 1);
+    if(!str) return NULL;
+    str[bits] = 0;
+
+    // type punning because signed shift is implementation-defined
+    unsigned u = *(unsigned *)&i;
+    for(; bits--; u >>= 1)
+        str[bits] = u & 1 ? '1' : '0';
+
+    return str;
+}
+
+static int interlogix_callback(bitbuffer_t *bitbuffer) {
+    char time_str[LOCAL_TIME_BUFLEN];
+    uint8_t *bb;
+    data_t *data;
+    unsigned int row = 0;
+    char device_type[2];
+    char *device_type_name = NULL;
+    char device_serial[7];
+    char device_raw_message[256];
+    char *low_battery = NULL;
+    char *f1_latch_state = NULL;
+    char *f2_latch_state = NULL;
+    char *f3_latch_state = NULL;
+    char *f4_latch_state = NULL;
+    char *f5_latch_state = NULL;
+
+    if (debug_output >= 1) 
+        fprintf(stderr,"GE/Interlogix Wireless Devices Template Callback\n");
+    
+    if (debug_output >= 2 ) 
+    {
+        fprintf(stderr, "bitbuffer:: Number of rows: %d \n", bitbuffer->num_rows);
+        fprintf(stderr, "bitbuffer:: Length of 1st row: %d \n", bitbuffer->bits_per_row[row]);
+        fprintf(stderr, "bitbuffer:: Printing 1st row: \n");
+
+        //this is typically done in the bitbuffer_print function but depends upon length so doing it explicitly here
+        for (unsigned int bit = 0; bit < bitbuffer->bits_per_row[row]; ++bit) {
+            if (bitbuffer->bb[row][bit/8] & (0x80 >> (bit % 8))) {
+                fprintf(stderr, "1");
+            } else {
+                fprintf(stderr, "0");
+            }
+        }
+        fprintf(stderr, "\n");
+        if (debug_output >=3)
+            //print all rows
+            bitbuffer_print(bitbuffer);
+    }
+
+    unsigned int bit_offset = bitbuffer_search(bitbuffer, row, 0, preamble, sizeof(preamble)*8);
+
+    if (bitbuffer->bits_per_row[row] - bit_offset < 45) //message should be at least 45 bits not including preamble bits
+    { 
+        if (debug_output >= 1)
+            fprintf(stderr, "Found valid preamble but message size too small, exiting! \n");
+        return 0;
+    }
+    else 
+    {
+        //set message starting postion (just past preamble and sync bit)
+        bit_offset = sizeof(preamble)*8 + bit_offset;
+
+        //even parity check (for bits 15-54) against even parity bit at bit 55
+        unsigned int calcParity = 0;
+        unsigned int msgParity = 0;
+        for (unsigned int bitc = bit_offset; bitc < bitbuffer->bits_per_row[row]; bitc++ )
+        {
+            if (bitc == (bit_offset+40)) { //15 bits for preamble and sync so bit 41 is actually documented bit 55
```

Comment:
wouldn't it be easier to just calculate the parity and then check the result after the loop?